lib/IDE/Refactoring.cpp

```diff
@@ -5397,6 +5404,149 @@ class AsyncConverter : private SourceEntityWalker {
     return true;
   }
 
+  /// Creates an async alternative function that forwards onto the completion
+  /// handler function through
+  /// withCheckedContinuation/withCheckedThrowingContinuation.
+  bool createAsyncWrapper() {
+    assert(Buffer.empty() && "AsyncConverter can only be used once");
+    auto *FD = cast<FuncDecl>(StartNode.get<Decl *>());
+
+    // First add the new async function declaration.
+    addFuncDecl(FD);
+    OS << tok::l_brace << "\n";
+
+    // Then add the body.
+    OS << tok::kw_return << " ";
+    if (TopHandler.HasError)
+      OS << tok::kw_try << " ";
+
+    OS << "await ";
+
+    // withChecked[Throwing]Continuation { cont in
+    if (TopHandler.HasError) {
+      OS << "withCheckedThrowingContinuation";
+    } else {
+      OS << "withCheckedContinuation";
+    }
+    OS << " " << tok::l_brace << " cont " << tok::kw_in << "\n";
+
+    // asyncFn(args...) {
+    addCallToAsyncMethod(FD, TopHandler);
```

Comment:
At some point it would be nice to handle the completion handler not being the last parameter, in which case this would fail. Could we add an assert here so we catch it if that does ever happen?

I think a comment here would help a lot as well - ie. this is actually calling the *sync* function, with the trailing closure added in the rest of the function.

Reply:
Decided to generalise the printing logic here to fit both use cases